code/game/say.dm

```diff
@@ -96,6 +96,21 @@ GLOBAL_LIST_INIT(freqtospan, list(
 	return "[say_mod(input, message_mode)][spanned ? ", \"[spanned]\"" : ""]"
 	// Citadel edit [spanned ? ", \"[spanned]\"" : ""]"
 
+/atom/movable/proc/say_emphasis(input)
+	input = encode_html_emphasis(input, "|", "i")
+	input = encode_html_emphasis(input, "_", "u")
+	input = encode_html_emphasis(input, "+", "b")
+	return input
+
+/**
+  * Replaces a character inside message with html tags.  Note that html var must not include brackets.
+  * Will not create an open html tag if it would not have a closing one.
+  */
+/atom/movable/proc/encode_html_emphasis(input, char, html)
+	var/regex/search = regex("([REGEX_QUOTE(char)])(\\b.*\\b)([REGEX_QUOTE(char)])", "g")
+	return search.Replace_char(input, "<[html]>$2</[html]>")
```

Comment:
```suggestion
#define ENCODE_HTML_EPHASIS(input, char, html, var) \
	var/static/regex/##var = regex("([REGEX_QUOTE(char)])(\\b[^[REGEX_QUOTE(char)]]+\\b)([REGEX_QUOTE(char)])", "g");\
	input = var.Replace_char(input, "<[html]>$2</[html]>")

/**
  * Replaces a character inside message with html tags.  Note that html var must not include brackets.
  * Will not create an open html tag if it would not have a closing one.
  */
/atom/movable/proc/say_emphasis(input)
	ENCODE_HTML_EPHASIS(input, "|", "i", italics)
	ENCODE_HTML_EPHASIS(input, "_", "u", underscore)
	ENCODE_HTML_EPHASIS(input, "+", "b", bold)
	return input

#undef ENCODE_HTML_EPHASIS
```

Speeding it up. Also changed the second group from `(\\b.*\\b)` to `(\\b[^REGEX_QUOTE(char)]+\\b)`, otherwise it'll count multiple markdowns of the same type as one.
(Make sure to apply the suggestion locally and correct possible typos first)